### DISCUSSION TITLE
Proposal: ref operator for a class:

### DISCUSSION THREAD

**Discussion by ritchiecarroll** [dotnet/csharplang#3687] [Language Ideas]
Create a `ref` operator for a `class`.

This would basically proxy `ref` return operations on a `class`, that would normally not allow this, to its ref operator.

Possible simpler solution to related dotnet/roslyn#12364 :

For example:

```c#
public sealed class MyClass<T>
{
    private T m_value;

    public MyClass(in T value) => m_value = value;

    public ref T Value => ref m_value;

    // Example of new proposed "ref" return operator for class:
    public static ref T operator ref(MyClass<T> value) => ref value.m_value;
}
```

Now code like this:
```c#
    var x = new MyClass<int>(12);
    ref var y = ref x.Value;
    y = 13;
```

Could become:
```c#
    var x = new MyClass<int>(12);
    ref var y = ref x;
    y = 13;
```

This would also allow inferred values like the following:

```c#
public struct MyStruct
{
    public int X;
}

var s = new MyClass<MyStruct>(new MyStruct { X = 12 });

// So instead of this:
s.Value.X = 13;
Console.WriteLine(s.Value.X);

// You could write this, where ref value is inferred from ref operator:
(ref s).X = 13;
Console.WriteLine((ref s).X);
```
I image for above, the parenthesis might be required to distinguish possibly ambiguities for ref property accessors and ref operator.  Otherwise you could do something _really_ fun and use a pointer operator (this gets my vote):
```C#
public sealed class MyClass<T>
{
    private T m_value;

    public MyClass(in T value) => m_value = value;

    public ref T Value => ref m_value;

    // Example of new proposed "ref" return operator for class using "*" operator:
    public static ref T operator *(MyClass<T> value) => ref value.m_value;
}

*s.X = 13;
Console.WriteLine(*s.X);
```
Should be easy enough to detect that this is not a multiply operator since there is a single parameter.

---

This kind of operator would allow results for large structures by doing something like the following:
```c#
public unsafe struct Matrix4x4
{
    public fixed double Values[16];
    public ref double this[int x, int y] => ref Values[x * y];
}

public class MatrixMultiplyOp
{
    private Matrix4x4 m_result;

    public MatrixMultiplyOp(in Matrix4x4 lhs, in Matrix4x4 rhs)
    {
        for (int i = 0; i < 4; i++)
            for (int j = 0; j < 4; j++)
                m_result[i, j] = lhs[i, j] * rhs[i, j];
    }

    public ref Matrix4x4 Result => ref m_result;

    // Example of new proposed "ref" return operator for class:
    public static ref Matrix4x4 operator *(MatrixMultiplyOp value) => ref value.m_result;
}

public static class Matrix
{
    public static MatrixMultiplyOp Multiply(in Matrix4x4 lhs, in Matrix4x4 rhs) => new MatrixMultiplyOp(lhs, rhs);

    // FYI - something like this already works today for this example, but the above code is more simple
    //public static ref Matrix4x4 Multiply(in Matrix4x4 lhs, in Matrix4x4 rhs) => ref new MatrixMultiplyOp(lhs, rhs).Result;
}

static void Main()
{
    // Used like this:
    var lhs = new Matrix4x4();
    var rhs = new Matrix4x4();
    ref var result = *Matrix.Multiply(lhs, rhs);

    // Would this infer "*" operator?
    ref var result =  ref Matrix.Multiply(lhs, rhs);
}
```

_Originally posted by @ritchiecarroll in https://github.com/dotnet/roslyn/issues/12364#issuecomment-656813277_

**Comment by ritchiecarroll** [comment]
Target use case: https://github.com/GridProtectionAlliance/go2cs/blob/master/src/gocore/golib/ptr.cs

**Comment by juliusfriedman** [comment]
Just use a Span?

```c#
using System;
public class C {
    ref struct Holder<T>{
Span<byte> bytes;
unsafe public Holder(ref T t){
//If you wanted to keep pointer here you could also
bytes= System.Runtime.InteropServices.MemoryMarshal.CreateSpan(ref System.Runtime.CompilerServices.Unsafe.As<T, byte>(ref t),  System.Runtime.CompilerServices.Unsafe.SizeOf<T>());
}

public T Data => System.Runtime.InteropServices.MemoryMarshal.CreateSpan(ref System.Runtime.CompilerServices.Unsafe.As<byte, T>(ref bytes[0]), 1)[0];

}
}
```

Since you can already pass a class by ref just use that and pass that by ref and then access by ref?

Am I misunderstanding something besides the potential type safety this would additionally provide? It's not called or spelled out as far as I can see.

You can also fix the pointer and just use unsafe... during a relation the stack is not moving anyway but I can see a benefit to this if it's also allowed in structs etc. (even only if `ref`) just for the type safety

After reading the syntax again I was confused, it's the syntax I would expect. Sorry.

I also done see the need because you can have a `ref` property on a class today already....


```c#
using System;
public class C {
    int m_Int;
    public ref int Int => ref m_Int;
}
```

We also have `WeakRefererence<T>` which you can use and pass by ref.

This would NOT require changes to the GC afaik because this is already possible, what is new is the operator `ref`

**Comment by svick** [comment]
I don't understand, what makes `(ref s).X` so much better than `s.Value.X` that it's worth modifying the language? Especially considering that `ref s` already means something, and you're overloading it to do something completely different.

**Comment by juliusfriedman** [comment]
> I don't understand, what makes `(ref s).X` so much better than `s.Value.X` that it's worth modifying the language? Especially considering that `ref s` already means something, and you're overloading it to do something completely different.

or just `ref s.Value.X` for that... I think the semantic difference is the deference... although not shown in the example


Completely speculating but:

```c#
ref S s = ref s;
ref V Value = ref s.Value;
ref int x = ref V.X;
```
Would be the long form of `(ref s).X` somehow although I cannot see how.

The other valuable thing is that this lets you choose what happens when someone takes a ref to your class. I can't say i don't like it ...

I want the same thing for `null`
https://github.com/dotnet/csharplang/issues/3692

Just thinking why this should be restricted to classes?

**Comment by ritchiecarroll** [comment]
As you can already add a `ref` to properties, this was more about picking a target ref operation for a class instance.

Like @svick mentions, there is not a tremendous difference by using `ref s.x` (where s is a class instance) vs `s.Value.x`.

However, the key was simply about providing a target `ref operator` for a class, denoting that the returned value would be a ref output. Note that no other operators support ref operations at this time (see dotnet/roslyn#12364) .

My thinking was at least with a `ref operator` for class, this could simplify interacting with large struct operations.

In the end, it is perhaps more of a convenience method to produce cleaner code - that said, isn't that the purpose of any operator? For example consider a class where primary operations involve interactions with a value of type `T` which could be a large `struct`. Then suppose type `T` is a large structure and has a property called `Altitude` you wish to modify. So if the class had the proposed  `ref` operator like the following:
```c#
     public static ref T operator ref(MyClass<T> value) => ref value.m_value;
```
Then instead of using a `ref` property to modify the `Altitude` value:
```c#
    myClass.Value.Altitude = 999;
```
You could instead do this:
```c#
    ref myClass.Altitude = 999;
```
If `MyClass<T>` already had an `Altitude` property, only then would you need to disambiguate:
```c#
    (ref myClass).Altitude = 999;
```

Again the key notion is to "proxy" a `ref` style operation for a class, basically promoting a single `ref` value for a class instance.

---
I also still think this could just as easily use "pointer deref" syntax instead of "ref" - same idea though:
```c#
    *myClass.Altitude = 999;
```
